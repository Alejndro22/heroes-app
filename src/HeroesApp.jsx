import { Outlet } from 'react-router';
import { Navbar } from './ui';

export const HeroesApp = () => {
  return (
    <>
      <Navbar />
      <Outlet />
    </>
  );
};
