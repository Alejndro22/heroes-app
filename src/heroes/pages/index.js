export * from './DcPage';
export * from './ErrorPage';
export * from './HeroPage';
export * from './MarvelPage';
export * from './SearchPage';
