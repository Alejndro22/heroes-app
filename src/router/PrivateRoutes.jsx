import { useContext, useEffect } from 'react';
import { Navigate, useLocation } from 'react-router-dom';
import { AuthContext } from '../auth';

// eslint-disable-next-line react/prop-types
export const PrivateRoutes = ({ children }) => {
  const { logged } = useContext(AuthContext);

  // useLocation hook returns the current location elements
  const { pathname, search } = useLocation();

  useEffect(() => {
    const lastPath = pathname + search;
    localStorage.setItem('lastPath', lastPath);
  }, [pathname, search]);

  if (pathname + search === '/') return <Navigate to={'/marvel'} />;

  return logged ? children : <Navigate to={'/login'} />;
};
