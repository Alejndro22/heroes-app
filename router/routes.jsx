import { HeroesApp } from '../src/HeroesApp';
import { LoginPage } from '../src/auth/pages/LoginPage';
import { DcPage } from '../src/heroes/pages/DcPage';
import { MarvelPage } from '../src/heroes/pages/MarvelPage';
import ErrorPage from '../src/heroes/pages/ErrorPage';

export const routes = [
  {
    path: '/',
    element: <HeroesApp />,
    errorElement: <ErrorPage />,
    children: [
      {
        path: 'marvel',
        element: <MarvelPage />,
      },
      {
        path: 'dc',
        element: <DcPage />,
      },
      {
        path: 'login',
        element: <LoginPage />,
      },
    ],
  },
];
