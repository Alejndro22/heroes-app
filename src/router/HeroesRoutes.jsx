// import { Navigate } from 'react-router-dom';
import { DcPage, HeroPage, MarvelPage, SearchPage } from '../heroes';

export const heroesChildrenRoutes = [
  {
    path: 'marvel',
    element: <MarvelPage />,
  },
  {
    path: 'dc',
    element: <DcPage />,
  },
  {
    path: 'search',
    element: <SearchPage />,
  },
  {
    path: 'hero/:id',
    element: <HeroPage />,
  },
  // {
  //   path: '/*',
  //   element: <Navigate to={'/marvel'} replace={true} />,
  // },
];
