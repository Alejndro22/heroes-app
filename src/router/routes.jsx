import { HeroesApp } from '../HeroesApp';
import { LoginPage } from '../auth';
import { DcPage, MarvelPage } from '../heroes';
import ErrorPage from '../heroes/pages/ErrorPage';

export const routes = [
  {
    path: '/',
    element: <HeroesApp />,
    errorElement: <ErrorPage />,
    children: [
      {
        path: 'marvel',
        element: <MarvelPage />,
      },
      {
        path: 'dc',
        element: <DcPage />,
      },
      {
        path: 'login',
        element: <LoginPage />,
      },
    ],
  },
];
