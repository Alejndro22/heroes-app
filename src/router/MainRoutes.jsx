import { HeroesRoutes } from '../heroes';
import { LoginPage } from '../auth';
import { heroesChildrenRoutes } from './HeroesRoutes';
import ErrorPage from '../heroes/pages/ErrorPage';
import { PrivateRoutes } from './PrivateRoutes';
import { PublicRoutes } from './PublicRoutes';

export const routes = [
  {
    path: '/login',
    element: (
      <PublicRoutes>
        <LoginPage />
      </PublicRoutes>
    ),
  },

  {
    path: '/',
    element: (
      <PrivateRoutes>
        <HeroesRoutes />
      </PrivateRoutes>
    ),
    errorElement: <ErrorPage />,
    children: heroesChildrenRoutes,
  },

  // For reference, this works when there's no need to check private/public routes
  // { path: 'login', element: <LoginPage /> },
];
