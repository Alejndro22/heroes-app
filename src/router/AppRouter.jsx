import { createBrowserRouter, RouterProvider } from 'react-router-dom';
import { routes } from './MainRoutes';

const router = createBrowserRouter(routes);

export const AppRouter = () => {
  return <RouterProvider router={router} />;
};
