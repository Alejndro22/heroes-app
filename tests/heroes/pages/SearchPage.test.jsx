import { fireEvent, render, screen } from '@testing-library/react';
import { routes } from '../../../src/router/MainRoutes';
import { SearchPage } from '../../../src/heroes';
import {
  MemoryRouter,
  RouterProvider,
  createMemoryRouter,
} from 'react-router-dom';
import { AuthContext } from '../../../src/auth';

// Mock useNavigate hook
const mockedUseNavigate = jest.fn();
// Here i tell to keep everything as default, and only useNavigate() will be mocked
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate,
}));

describe('tests on <SearchPage />', () => {
  beforeEach(() => jest.clearAllMocks());

  const contextValue = {
    logged: true,
    user: { id: 'abc', name: 'testing' },
  };

  test('should match snapshot with default values', () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
    // screen.debug();
  });

  test('should display desired value, and the input should be the value of querystring', () => {
    const router = createMemoryRouter(routes, {
      initialEntries: ['/search?q=batman'],
    });

    render(
      <AuthContext.Provider value={contextValue}>
        <RouterProvider router={router}>
          <SearchPage />
        </RouterProvider>
      </AuthContext.Provider>
    );

    const input = screen.getByRole('textbox');
    expect(input.value).toBe('batman');

    //to make sure that card is being displayed, here I check if the image src is correct
    const img = screen.getByRole('img');
    expect(img.src).toContain('/assets/heroes/dc-batman.jpg');
  });

  test('should display error if not hero found', () => {
    const router = createMemoryRouter(routes, {
      initialEntries: ['/search?q=batman123'],
    });

    render(
      <AuthContext.Provider value={contextValue}>
        <RouterProvider router={router}>
          <SearchPage />
        </RouterProvider>
      </AuthContext.Provider>
    );

    expect(screen.getByText('No hero found with q=batman123')).toBeTruthy();
  });

  test('should call Navigate to the new path', () => {
    const inputValue = 'testingSearch';

    const router = createMemoryRouter(routes, {
      initialEntries: ['/search?q=batman123'],
    });

    render(
      <AuthContext.Provider value={contextValue}>
        <RouterProvider router={router}>
          <SearchPage />
        </RouterProvider>
      </AuthContext.Provider>
    );

    const input = screen.getByRole('textbox');
    fireEvent.input(input, { target: { value: inputValue } });

    const form = screen.getByLabelText('form-test');
    fireEvent.submit(form);

    expect(mockedUseNavigate).toHaveBeenCalledWith(`?q=${inputValue}`);
  });
});
