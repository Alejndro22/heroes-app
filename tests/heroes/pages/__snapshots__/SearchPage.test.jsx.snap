// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on <SearchPage /> should match snapshot with default values 1`] = `
<div>
  <h1
    class="mb-4 text-4xl font-bold leading-none tracking-tight text-gray-900 md:text-5xl lg:text-6xl"
  >
    Search Page
  </h1>
  <hr />
  <div
    class="sm:grid sm:grid-cols-12 sm:gap-12"
  >
    <div
      class="col-span-5 space-y-2"
    >
      <h4
        class="mt-3 text-lg font-semibold text-gray-900 md:text-xl lg:text-2xl"
      >
        Searching
      </h4>
      <hr />
      <form
        aria-label="form-test"
        class="flex flex-col"
      >
        <input
          autocomplete="off"
          class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-munsell-blue focus:border-munsell-blue block w-full p-2.5"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="mt-4 bg-transparent text-munsell-blue hover:bg-munsell-blue hover:text-white font-semibold py-2 px-4 border border-munsell-blue hover:border-transparent rounded w-min"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-span-7 space-y-2"
    >
      <h4
        class="mt-3 text-lg font-semibold text-gray-900 md:text-xl lg:text-2xl"
      >
        Results
      </h4>
      <hr />
      <div
        class="flex items-center bg-munsell-blue text-white text-sm font-bold px-4 py-3 bg-opacity-80 animate__animated animate__fadeIn"
        role="alert"
      >
        <svg
          class="fill-current w-4 h-4 mr-2"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.432 0c1.34 0 2.01.912 2.01 1.957 0 1.305-1.164 2.512-2.679 2.512-1.269 0-2.009-.75-1.974-1.99C9.789 1.436 10.67 0 12.432 0zM8.309 20c-1.058 0-1.833-.652-1.093-3.524l1.214-5.092c.211-.814.246-1.141 0-1.141-.317 0-1.689.562-2.502 1.117l-.528-.88c2.572-2.186 5.531-3.467 6.801-3.467 1.057 0 1.233 1.273.705 3.23l-1.391 5.352c-.246.945-.141 1.271.106 1.271.317 0 1.357-.392 2.379-1.207l.6.814C12.098 19.02 9.365 20 8.309 20z"
          />
        </svg>
        <p>
          Search a hero
        </p>
      </div>
      <div
        class="flex flex-col items-center space-y-2"
      />
    </div>
  </div>
</div>
`;
